File: src/option_tbn/tablet.js
import path from "path";
import { ReadableStream } from "@swimburger/isomorphic-streams";
import { isEmpty, chunksToLines } from "../stream.js";
import { optionLine } from "./line.js";

export async function optionTabletStream(fs, dir, tablet, { query, matchMap }) {
  const filepath = path.join(dir, tablet.filename);

  const lineStream = (await isEmpty(fs, filepath))
    ? ReadableStream.from([])
    : chunksToLines(fs.createReadStream(filepath));

  const lineIterator = lineStream[Symbol.asyncIterator]();

  const entryInitial = { _: tablet.base };

  let stateSaved = {
    entry: entryInitial,
    fst: undefined,
    isMatch: false,
    matchMap,
  };

  async function pullLine(state) {
    const { done, value } = await lineIterator.next();

    if (done) {
      return { done: true, value: state };
    }

    const stateLine = optionLine(tablet, state, value);

    if (stateLine.last) {
      return { done: false, value: stateLine };
    } else {
      return pullLine(stateLine);
    }
  }

  return new ReadableStream({
    async pull(controller) {
      const { done, value } = await pullLine(stateSaved);

      if (done) {
        if (value.isMatch) {
          controller.enqueue(value);
        }

        controller.close();
      }

      if (value.last) {
        controller.enqueue(value.last);

        value.last = false;
      }

      stateSaved = value;
    },
  });
}

File: src/dataset/option/tablet.rs
use super::line::{option_line, State};
use super::strategy::Tablet;
use crate::{line::Line, Dataset, Entry, Result};
use async_stream::try_stream;
use futures_core::stream::Stream;
use std::fs::File;

pub fn option_tablet_stream(
    dataset: Dataset,
    tablet: Tablet,
    state: State,
) -> impl Stream<Item = Result<State>> {
    try_stream! {
        let filepath = dataset.dir.join(&tablet.filename);

        if std::fs::metadata(&filepath).is_err() {
            return;
        }

        let mut rdr = csv::ReaderBuilder::new()
            .has_headers(false)
            .flexible(true)
            .from_reader(File::open(&filepath)?);

        let mut state = State {
            query: state.query.clone(),
            entry: Entry::new(&tablet.base),
            fst: None,
            is_match: false,
            match_map: state.match_map.clone(),
            last: None,
        };

        for result in rdr.records() {
            let record = result?;

            let line_escaped = Line {
                key: match record.get(0) { None => String::from(""), Some(s) => s.to_owned() },
                value: match record.get(1) { None => String::from(""), Some(s) => s.to_owned() }
            };

            let line = line_escaped.unescape();

            state = option_line(tablet.clone(), state, line)?;

            if state.last.is_some()  {
                yield state.clone();

                state.last = None;
            }
        }

        if state.is_match {
            state.last = Some(state.clone().entry);

            yield state;
        }
    }
}
